### PR TITLE
chore: check each feature works properly

### LINKED ISSUE #1693: tokio "fs" feature should depend on tokio-io
With the following `Cargo.toml` dependency (latest [HEAD](https://github.com/tokio-rs/tokio/commit/227533d456fe32e48ffcd3796f1e6c8f9318b230) at the moment of creating this issue)

```toml
[dependencies.tokio]
git = "https://github.com/tokio-rs/tokio.git"
rev = "227533d456fe32e48ffcd3796f1e6c8f9318b230"
default-features = false
features = ["fs"]
```

compilation fails because fs routines are using `tokio-io` sub-crate, which is excluded by features selection:

```
$ cargo +nightly check
    Checking tokio v0.2.0-alpha.6 (https://github.com/tokio-rs/tokio.git?rev=227533d456fe32e48ffcd3796f1e6c8f9318b230#227533d4)
error[E0432]: unresolved import `tokio_io`
 --> /home/svartalf/.cargo/git/checkouts/tokio-377c595163f99a10/227533d/tokio/src/fs/file.rs:9:5
  |
9 | use tokio_io::{AsyncRead, AsyncWrite};
  |     ^^^^^^^^ use of undeclared type or module `tokio_io`

error[E0432]: unresolved import `tokio_io`
 --> /home/svartalf/.cargo/git/checkouts/tokio-377c595163f99a10/227533d/tokio/src/fs/blocking.rs:3:5
  |
3 | use tokio_io::{AsyncRead, AsyncWrite};
  |     ^^^^^^^^ use of undeclared type or module `tokio_io`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0432`.
error: could not compile `tokio`.
```

I suppose that `fs` feature should enable `io` feature too?

### PR DESCRIPTION
**Pull request by taiki-e**
The current tests for each feature gate do not work properly because the default feature is enabled by dev-dependencies. 

The cause of this is https://github.com/rust-lang/cargo/issues/4866, but it seems difficult to fix on the cargo side.
~~So remove all dev-dependencies in CI to avoid this issue. This is basically the same as https://github.com/rust-lang-nursery/futures-rs/pull/1456#issuecomment-465312563.~~ EDIT2 see https://github.com/tokio-rs/tokio/pull/1695#issuecomment-547875992

~~Also, this change test to run only once with all-features instead of running tests for each feature.~~ EDIT: see https://github.com/tokio-rs/tokio/pull/1695#issuecomment-546615264

Closes #1693 

Refs: 
  * https://github.com/rust-lang/cargo/issues/4866
  * https://github.com/rust-lang-nursery/futures-rs/issues/1396
  * https://github.com/rust-lang-nursery/futures-rs/pull/1456

### CODE REVIEW

**Comment by taiki-e** [comment]
> Also, this change test to run only once with all-features instead of running tests for each feature.

Oh, this is not good. I also need to keep the existing one for tests-build.

**Comment by taiki-e** [comment]
hmm...
```text
error[E0432]: unresolved import `tokio_sync`
 --> tokio/src/net/driver/reactor.rs:4:5
  |
4 | use tokio_sync::AtomicWaker;
  |     ^^^^^^^^^^ use of undeclared type or module `tokio_sync`

error[E0432]: unresolved import `slab`
 --> tokio/src/net/driver/reactor.rs:7:5
  |
7 | use slab::Slab;
  |     ^^^^ help: a similar path exists: `mio::slab`

error[E0432]: unresolved import `crate::io`
 --> tokio/src/net/util/poll_evented.rs:1:12
  |
1 | use crate::io::{AsyncRead, AsyncWrite};
  |            ^^
  |            |
  |            unresolved import
  |            help: a similar path exists: `std::io`
```

**Comment by carllerche** [comment]
Thanks. I haven't reviewed yet. However, I want to quickly point out that `tests-build` exists to work around the cargo limitations and test the various cargo feature combos. These tests have gone out of date w/ the recent churn though.

**Comment by taiki-e** [comment]
> However, I want to quickly point out that `tests-build` exists to work around the cargo limitations and test the various cargo feature combos. These tests have gone out of date w/ the recent churn though.

`tests-build`'s approach requires all tests to be written manually. Also, we need to write tests for each platform for some crates (see https://github.com/tokio-rs/tokio/pull/1695/commits/7aee1f85f691f7e125f11e6f7cca4b2b1b04a074).

But I'd like to try `tests-build` as another pr because it seems help reduce CI time.

**Comment by carllerche** [comment]
@taiki-e I wasn't making a comment on your PR yet, just providing context :) I will probably review on monday.

**Comment by taiki-e** [comment]
UPDATE: It is very hard to maintain these tests manually, so I'm writing a tool to make these checks run in one line :)

**Comment by carllerche** [comment]
:+1: Should I merge this then or wait?

**Comment by carllerche** [comment]
Also, is the plan to get rid of `tests-build` or have something else with it?

**Comment by taiki-e** [comment]
I wrote https://github.com/taiki-e/cargo-hack to improve feature flag testing.
cargo-hack is basically wrapper of cargo that propagates subcommand and provides additional flags to avoid some of the limitations of cargo. e.g.:
  * --each-feature - run for each feature which includes default features and --no-default-features (if the crate has one or more feature) of package
  * --no-dev-deps - run without dev-dependencies to avoid https://github.com/rust-lang/cargo/issues/4866

By using these two features, you can easily test the feature flag gate: 
```text
$ cargo hack check --each-feature --no-dev-deps
```

<details>
<summary> output: </summary>

("info: running `cargo check ...` on tokio" shows what command is actually being executed.)

```text
info: running `cargo check` on tokio
    Checking tokio v0.2.0-alpha.6 (/home/vsts/work/1/s/tokio)
    Finished dev [unoptimized + debuginfo] target(s) in 23.38s
info: running `cargo check --no-default-features` on tokio
    Checking tokio v0.2.0-alpha.6 (/home/vsts/work/1/s/tokio)
    Finished dev [unoptimized + debuginfo] target(s) in 0.10s
info: running `cargo check --features=executor-core --no-default-features` on tokio
    Checking tokio v0.2.0-alpha.6 (/home/vsts/work/1/s/tokio)
    Finished dev [unoptimized + debuginfo] target(s) in 0.31s
info: running `cargo check --features=blocking --no-default-features` on tokio
    Checking tokio v0.2.0-alpha.6 (/home/vsts/work/1/s/tokio)
    Finished dev [unoptimized + debuginfo] target(s) in 1.13s
info: running `cargo check --features=fs --no-default-features` on tokio
    Checking tokio v0.2.0-alpha.6 (/home/vsts/work/1/s/tokio)
    Finished dev [unoptimized + debuginfo] target(s) in 1.67s
info: running `cargo check --features=io-traits --no-default-features` on tokio
    Checking tokio v0.2.0-alpha.6 (/home/vsts/work/1/s/tokio)
    Finished dev [unoptimized + debuginfo] target(s) in 0.24s
info: running `cargo check --features=io-util --no-default-features` on tokio
    Checking tokio v0.2.0-alpha.6 (/home/vsts/work/1/s/tokio)
    Finished dev [unoptimized + debuginfo] target(s) in 0.79s
info: running `cargo check --features=io --no-default-features` on tokio
    Checking tokio v0.2.0-alpha.6 (/home/vsts/work/1/s/tokio)
    Finished dev [unoptimized + debuginfo] target(s) in 0.81s
info: running `cargo check --features=macros --no-default-features` on tokio
   Compiling proc-macro2 v1.0.6
   Compiling syn v1.0.7
   Compiling quote v1.0.2
   Compiling tokio-macros v0.2.0-alpha.6 (/home/vsts/work/1/s/tokio-macros)
    Checking tokio v0.2.0-alpha.6 (/home/vsts/work/1/s/tokio)
    Finished dev [unoptimized + debuginfo] target(s) in 12.47s
info: running `cargo check --features=net-full --no-default-features` on tokio
    Checking tokio v0.2.0-alpha.6 (/home/vsts/work/1/s/tokio)
    Finished dev [unoptimized + debuginfo] target(s) in 2.76s
info: running `cargo check --features=net-driver --no-default-features` on tokio
    Checking tokio v0.2.0-alpha.6 (/home/vsts/work/1/s/tokio)
    Finished dev [unoptimized + debuginfo] target(s) in 1.73s
info: running `cargo check --features=rt-current-thread --no-default-features` on tokio
    Checking tokio v0.2.0-alpha.6 (/home/vsts/work/1/s/tokio)
    Finished dev [unoptimized + debuginfo] target(s) in 2.43s
info: running `cargo check --features=rt-full --no-default-features` on tokio
    Checking tokio v0.2.0-alpha.6 (/home/vsts/work/1/s/tokio)
    Finished dev [unoptimized + debuginfo] target(s) in 4.50s
info: running `cargo check --features=signal --no-default-features` on tokio
    Checking tokio v0.2.0-alpha.6 (/home/vsts/work/1/s/tokio)
    Finished dev [unoptimized + debuginfo] target(s) in 1.79s
info: running `cargo check --features=sync --no-default-features` on tokio
    Checking tokio v0.2.0-alpha.6 (/home/vsts/work/1/s/tokio)
    Finished dev [unoptimized + debuginfo] target(s) in 0.94s
info: running `cargo check --features=tcp --no-default-features` on tokio
    Checking tokio v0.2.0-alpha.6 (/home/vsts/work/1/s/tokio)
    Finished dev [unoptimized + debuginfo] target(s) in 2.45s
info: running `cargo check --features=timer --no-default-features` on tokio
    Checking tokio v0.2.0-alpha.6 (/home/vsts/work/1/s/tokio)
    Finished dev [unoptimized + debuginfo] target(s) in 1.60s
info: running `cargo check --features=udp --no-default-features` on tokio
    Checking tokio v0.2.0-alpha.6 (/home/vsts/work/1/s/tokio)
    Finished dev [unoptimized + debuginfo] target(s) in 2.41s
info: running `cargo check --features=uds --no-default-features` on tokio
    Checking tokio v0.2.0-alpha.6 (/home/vsts/work/1/s/tokio)
    Finished dev [unoptimized + debuginfo] target(s) in 2.53s
info: running `cargo check --features=process --no-default-features` on tokio
    Checking tokio v0.2.0-alpha.6 (/home/vsts/work/1/s/tokio)
    Finished dev [unoptimized + debuginfo] target(s) in 2.70s
```

</details>

This will remove the need to manually list feature flags and tests-build should no longer be needed.

File: azure-pipelines.yml [hunk 1]
```diff
@@ -47,29 +35,10 @@ jobs:
     displayName: Test sub crates -
     rust: beta
     crates:
-      tokio-macros: []
-      tokio-test: []
-      tokio-util: []
-      examples: []
-
-# Test compilation failure
-# Disable pending: https://github.com/tokio-rs/tokio/pull/1695#issuecomment-547045383
```
Comment (taiki-e) [inline azure-pipelines.yml]: Removed :)

File: ci/azure-test-stable.yml [hunk 2]
```diff
@@ -48,14 +67,25 @@ jobs:
...
+    # * --each-feature - run for each feature which includes --no-default-features and default features of package
+    # * --no-dev-deps - build without dev-dependencies to avoid https://github.com/rust-lang/cargo/issues/4866
+    - script: cargo hack check --each-feature --no-dev-deps
+      env:
+        LOOM_MAX_PREEMPTIONS: 2
+        CI: 'True'
+      displayName: ${{ crate }} - cargo hack check --each-feature
+      workingDirectory: $(Build.SourcesDirectory)/${{ crate }}
+      # FIXME(taiki-e): many features of current cargo-hack are not available on windows: https://github.com/taiki-e/cargo-hack/issues/3
+      condition: not(eq(variables['Agent.OS'], 'Windows_NT'))
```
Comment (taiki-e) [inline ci/azure-test-stable.yml]: Unfortunately, many features of current cargo-hack are not available on windows. (https://github.com/taiki-e/cargo-hack/issues/3)

Comment (taiki-e) [reply ci/azure-test-stable.yml]: However, it seems many features are not directly related to OS-specific dependencies (https://github.com/tokio-rs/tokio/pull/1695/commits/7aee1f85f691f7e125f11e6f7cca4b2b1b04a074), so for now, I think this is almost all right.

**Comment by taiki-e** [comment]
BTW, another benefit is that CI time is significantly reduced.
* this - windows/linux: 13m-14m  macos: 11m-12m (https://dev.azure.com/tokio-rs/Tokio/_build/results?buildId=3030)
* previous - windows/linux: 40m-50m macos: 30m-40m (https://dev.azure.com/tokio-rs/Tokio/_build/results?buildId=3020)

**Comment by taiki-e** [comment]
@carllerche Updated PR to use cargo-hack. Could you review it again?

File: ci/azure-test-stable.yml (hunk 2 above)
Comment (taiki-e) [reply ci/azure-test-stable.yml]: I tried to fix this in https://github.com/taiki-e/cargo-hack/pull/6, but I couldn't fix it completely.

File: ci/azure-test-stable.yml [hunk 3]
```diff
@@ -20,6 +20,11 @@ jobs:
       # rust_version: stable
       rust_version: ${{ parameters.rust }}
 
+  - script: cargo install cargo-hack
```
Comment (hawkw) [inline ci/azure-test-stable.yml]: This is nitpicky, but it might be helpful to have a comment here explaining what we're using `cargo-hack` for and why?

Comment (taiki-e) [reply ci/azure-test-stable.yml]: I wrote descriptions for 'check' scripts, but if necessary, I'd happy to add a description to the 'install' script.
https://github.com/tokio-rs/tokio/blob/02f7264008e0333d7c4df183d3bcfbda8a6d930f/ci/azure-test-stable.yml#L47-L50
https://github.com/tokio-rs/tokio/blob/02f7264008e0333d7c4df183d3bcfbda8a6d930f/ci/azure-test-stable.yml#L78-L81

**Comment by taiki-e** [comment]
@carllerche @hawkw Thanks for the review!

**Pull request merged**